lib/ansible/modules/cloud/amazon/kms_facts.py

```diff
@@ -0,0 +1,313 @@
+            kwargs['Marker'] = tag_response['NextMarker']
+        else:
+            more = False
+    return boto3_tag_list_to_ansible_dict(tags)
+
+
+def key_matches_filter(key, filtr):
+    if filtr[0] == 'key-id':
+        return filtr[1] == key['KeyId']
+    if filtr[0] == 'tag-key':
+        return filtr[1] in key['Tags']
+    if filtr[0] == 'tag-value':
+        return filtr[1] in key['Tags'].values()
+    if filtr[0].startswith('tag:'):
+        return key['Tags'][filtr[0][4:]] == filtr[1]
+
+
+def key_matches_filters(key, filters):
+    if not filters:
+        return True
+    else:
+        return all([key_matches_filter(key, filtr) for filtr in filters.items()])
+
+
+def get_key_details(connection, module, key_id, tokens=[]):
+    result = dict()
+    try:
+        metadata = get_kms_metadata_with_backoff(connection, key_id)['KeyMetadata']
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed to obtain key metadata",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+    del(metadata['KeyId'])
+    del(metadata['Arn'])
+    result.update(metadata)
+
+    result['Tags'] = get_kms_tags(connection, module, key_id)
+    try:
+        result['Grants'] = get_kms_grants_with_backoff(connection, key_id, tokens=tokens)['Grants']
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed to obtain key grants",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+    return result
+
+
+def get_kms_facts(connection, module):
+    try:
+        keys = get_kms_keys_with_backoff(connection)['Keys']
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed to obtain keys",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+
+    for key in keys:
+        key.update(get_key_details(connection, module, key['KeyId']))
+
+    return keys
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            filters=dict(type='dict')
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 and botocore are required for this module')
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    if region:
+        connection = boto3_conn(module, conn_type='client', resource='kms', region=region, endpoint=ec2_url, **aws_connect_params)
+    else:
+        module.fail_json(msg="region must be specified")
+
+    all_keys = get_kms_facts(connection, module)
+    filtered = [key for key in all_keys if key_matches_filters(key, module.params['filters'])]
+    module.exit_json(keys=[camel_dict_to_snake_dict(key) for key in filtered])
```

Comment:
You've done more testing on the tag functionality than I have, sorry!

It's always best to do camel_to_snake before doing tag conversion, I should know this by now.